File: packages/server/backbone/go/src/cache.go
package main

import (
	"database/sql"
	"log"
	"sync"
	"time"
)

type CacheStore struct {
	mu   sync.RWMutex
	data map[string]*CacheEntry
	db   *sql.DB
}

func NewCacheStore(db *sql.DB) *CacheStore {
	cs := &CacheStore{
		data: make(map[string]*CacheEntry),
		db:   db,
	}
	cs.loadFromDB()
	go cs.evictLoop()
	return cs
}

func (cs *CacheStore) loadFromDB() {
	rows, err := cs.db.Query(`SELECT key, value, ttl, expires_at, created_at, updated_at FROM _cache`)
	if err != nil {
		log.Printf("cache: load from db: %v", err)
		return
	}
	defer rows.Close()
	for rows.Next() {
		var e CacheEntry
		if err := rows.Scan(&e.Key, &e.Value, &e.TTL, &e.ExpiresAt, &e.CreatedAt, &e.UpdatedAt); err != nil {
			log.Printf("cache: scan row: %v", err)
			continue
		}
		if e.TTL > 0 && e.ExpiresAt != "" {
			expires, err := time.Parse(time.RFC3339, e.ExpiresAt)
			if err == nil && time.Now().UTC().After(expires) {
				continue
			}
		}
		cs.data[e.Key] = &e
	}
}

func (cs *CacheStore) evictLoop() {
	ticker := time.NewTicker(30 * time.Second)
	for range ticker.C {
		now := time.Now().UTC()
		cs.mu.Lock()
		for key, entry := range cs.data {
			if entry.TTL > 0 && entry.ExpiresAt != "" {
				expires, err := time.Parse(time.RFC3339, entry.ExpiresAt)
				if err == nil && now.After(expires) {
					delete(cs.data, key)
				}
			}
		}
		cs.mu.Unlock()
	}
}

func (cs *CacheStore) Set(key, value string, ttl int) *CacheEntry {
	now := time.Now().UTC()
	nowStr := now.Format(time.RFC3339)
	var expiresAt string
	if ttl > 0 {
		expiresAt = now.Add(time.Duration(ttl) * time.Second).Format(time.RFC3339)
	}
	entry := &CacheEntry{
		Key:       key,
		Value:     value,
		TTL:       ttl,
		ExpiresAt: expiresAt,
		CreatedAt: nowStr,
		UpdatedAt: nowStr,
	}
	cs.mu.Lock()
	cs.data[key] = entry
	cs.mu.Unlock()
	cs.persist(entry)
	return entry
}

func (cs *CacheStore) Get(key string) *CacheEntry {
	cs.mu.RLock()
	entry, ok := cs.data[key]
	cs.mu.RUnlock()
	if !ok {
		entry = cs.load(key)
		if entry == nil {
			return nil
		}
	}
	if entry.TTL > 0 && entry.ExpiresAt != "" {
		expires, err := time.Parse(time.RFC3339, entry.ExpiresAt)
		if err == nil && time.Now().UTC().After(expires) {
			cs.mu.Lock()
			delete(cs.data, key)
			cs.mu.Unlock()
			cs.db.Exec(`DELETE FROM _cache WHERE key = ?`, key)
			return nil
		}
	}
	return entry
}

func (cs *CacheStore) load(key string) *CacheEntry {
	var e CacheEntry
	err := cs.db.QueryRow(`SELECT key, value, ttl, expires_at, created_at, updated_at FROM _cache WHERE key = ?`, key).
		Scan(&e.Key, &e.Value, &e.TTL, &e.ExpiresAt, &e.CreatedAt, &e.UpdatedAt)
	if err != nil {
		return nil
	}
	if e.TTL > 0 && e.ExpiresAt != "" {
		expires, err := time.Parse(time.RFC3339, e.ExpiresAt)
		if err == nil && time.Now().UTC().After(expires) {
			cs.db.Exec(`DELETE FROM _cache WHERE key = ?`, key)
			return nil
		}
	}
	cs.mu.Lock()
	cs.data[key] = &e
	cs.mu.Unlock()
	return &e
}

func (cs *CacheStore) Delete(key string) bool {
	cs.mu.Lock()
	_, ok := cs.data[key]
	delete(cs.data, key)
	cs.mu.Unlock()
	cs.db.Exec(`DELETE FROM _cache WHERE key = ?`, key)
	return ok
}

func (cs *CacheStore) List() []CacheEntry {
	now := time.Now().UTC()
	cs.mu.RLock()
	entries := make([]CacheEntry, 0, len(cs.data))
	for _, e := range cs.data {
		if e.TTL > 0 && e.ExpiresAt != "" {
			expires, err := time.Parse(time.RFC3339, e.ExpiresAt)
			if err == nil && now.After(expires) {
				continue
			}
		}
		entries = append(entries, *e)
	}
	cs.mu.RUnlock()
	return entries
}

func (cs *CacheStore) Flush() {
	cs.mu.Lock()
	cs.data = make(map[string]*CacheEntry)
	cs.mu.Unlock()
	cs.db.Exec(`DELETE FROM _cache`)
}

func (cs *CacheStore) Stats() map[string]any {
	cs.mu.RLock()
	total := len(cs.data)
	now := time.Now().UTC()
	expired := 0
	for _, e := range cs.data {
		if e.TTL > 0 && e.ExpiresAt != "" {
			expires, err := time.Parse(time.RFC3339, e.ExpiresAt)
			if err == nil && now.After(expires) {
				expired++
			}
		}
	}
	cs.mu.RUnlock()
	return map[string]any{
		"total_entries":   total,
		"expired_entries": expired,
		"memory":          true,
	}
}

func (cs *CacheStore) persist(entry *CacheEntry) {
	cs.db.Exec(`INSERT INTO _cache (key, value, ttl, expires_at, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?)
		ON CONFLICT(key) DO UPDATE SET value = excluded.value, ttl = excluded.ttl, expires_at = excluded.expires_at, updated_at = excluded.updated_at`,
		entry.Key, entry.Value, entry.TTL, entry.ExpiresAt, entry.CreatedAt, entry.UpdatedAt)
}

File: packages/server/backbone/rust/src/cache.rs
use std::collections::HashMap;
use std::sync::RwLock;

use chrono::Utc;
use rusqlite::Connection;

use crate::db;
use crate::models::*;

pub struct CacheStore {
    pub data: RwLock<HashMap<String, CacheEntry>>,
}

impl CacheStore {
    pub fn new() -> Self {
        CacheStore {
            data: RwLock::new(HashMap::new()),
        }
    }

    pub fn load_from_db(&self, conn: &Connection) {
        if let Ok(entries) = db::list_cache_entries(conn) {
            let mut data = self.data.write().unwrap();
            for entry in entries {
                data.insert(entry.key.clone(), entry);
            }
        }
    }

    pub fn set(&self, conn: &Connection, key: &str, value: &str, ttl: i64) -> Result<CacheEntry> {
        let entry = db::set_cache_entry(conn, key, value, ttl)?;
        let mut data = self.data.write().unwrap();
        data.insert(key.to_string(), entry.clone());
        Ok(entry)
    }

    pub fn get(&self, conn: &Connection, key: &str) -> Result<Option<CacheEntry>> {
        {
            let data = self.data.read().unwrap();
            if let Some(entry) = data.get(key) {
                if entry.ttl > 0
                    && !entry.expires_at.is_empty()
                    && let Ok(expires) = chrono::DateTime::parse_from_rfc3339(&entry.expires_at)
                    && Utc::now() > expires
                {
                    drop(data);
                    let mut data = self.data.write().unwrap();
                    data.remove(key);
                    let _ = db::delete_cache_entry(conn, key);
                    return Ok(None);
                }
                return Ok(Some(entry.clone()));
            }
        }

        if let Some(entry) = db::get_cache_entry(conn, key)? {
            let mut data = self.data.write().unwrap();
            data.insert(key.to_string(), entry.clone());
            Ok(Some(entry))
        } else {
            Ok(None)
        }
    }

    pub fn delete(&self, conn: &Connection, key: &str) -> Result<bool> {
        let mut data = self.data.write().unwrap();
        data.remove(key);
        db::delete_cache_entry(conn, key)
    }

    pub fn list(&self) -> Vec<CacheEntry> {
        let data = self.data.read().unwrap();
        let now = Utc::now();
        data.values()
            .filter(|e| {
                if e.ttl > 0
                    && !e.expires_at.is_empty()
                    && let Ok(expires) = chrono::DateTime::parse_from_rfc3339(&e.expires_at)
                {
                    return now <= expires;
                }
                true
            })
            .cloned()
            .collect()
    }

    pub fn flush(&self, conn: &Connection) -> Result<()> {
        let mut data = self.data.write().unwrap();
        data.clear();
        db::flush_cache(conn)
    }

    pub fn stats(&self) -> serde_json::Value {
        let data = self.data.read().unwrap();
        let now = Utc::now();
        let total = data.len();
        let expired = data
            .values()
            .filter(|e| {
                if e.ttl > 0
                    && !e.expires_at.is_empty()
                    && let Ok(expires) = chrono::DateTime::parse_from_rfc3339(&e.expires_at)
                {
                    return now > expires;
                }
                false
            })
            .count();
        serde_json::json!({
            "total_entries": total,
            "expired_entries": expired,
            "memory": true,
        })
    }

    pub fn evict_expired(&self) {
        let mut data = self.data.write().unwrap();
        let now = Utc::now();
        data.retain(|_, e| {
            if e.ttl > 0
                && !e.expires_at.is_empty()
                && let Ok(expires) = chrono::DateTime::parse_from_rfc3339(&e.expires_at)
            {
                return now <= expires;
            }
            true
        });
    }
}

pub fn start_eviction_loop(cache: std::sync::Arc<CacheStore>) {
    tokio::spawn(async move {
        let mut interval = tokio::time::interval(std::time::Duration::from_secs(30));
        loop {
            interval.tick().await;
            cache.evict_expired();
        }
    });
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::{Duration, Utc};

    #[test]
    fn new_creates_empty_cache() {
        let cache = CacheStore::new();
        assert!(cache.data.read().unwrap().is_empty());
    }

    #[test]
    fn list_returns_empty_for_new_cache() {
        let cache = CacheStore::new();
        assert!(cache.list().is_empty());
    }

    #[test]
    fn stats_returns_zero_counts_for_new_cache() {
        let cache = CacheStore::new();
        let stats = cache.stats();
        assert_eq!(stats["total_entries"], 0);
        assert_eq!(stats["expired_entries"], 0);
    }

    #[test]
    fn list_filters_expired_entries() {
        let cache = CacheStore::new();
        let now = Utc::now();
        let past = now - Duration::hours(1);
        let future = now + Duration::hours(1);

        {
            let mut data = cache.data.write().unwrap();
            data.insert("expired".to_string(), CacheEntry {
                key: "expired".to_string(),
                value: "old".to_string(),
                ttl: 3600,
                expires_at: past.to_rfc3339(),
                created_at: past.to_rfc3339(),
                updated_at: past.to_rfc3339(),
            });
            data.insert("valid".to_string(), CacheEntry {
                key: "valid".to_string(),
                value: "fresh".to_string(),
                ttl: 3600,
                expires_at: future.to_rfc3339(),
                created_at: now.to_rfc3339(),
                updated_at: now.to_rfc3339(),
            });
        }

        let entries = cache.list();
        assert_eq!(entries.len(), 1);
        assert_eq!(entries[0].key, "valid");
    }

    #[test]
    fn evict_expired_removes_expired_entries() {
        let cache = CacheStore::new();
        let now = Utc::now();
        let past = now - Duration::hours(1);
        let future = now + Duration::hours(1);

        {
            let mut data = cache.data.write().unwrap();
            data.insert("expired".to_string(), CacheEntry {
                key: "expired".to_string(),
                value: "old".to_string(),
                ttl: 3600,
                expires_at: past.to_rfc3339(),
                created_at: past.to_rfc3339(),
                updated_at: past.to_rfc3339(),
            });
            data.insert("valid".to_string(), CacheEntry {
                key: "valid".to_string(),
                value: "fresh".to_string(),
                ttl: 3600,
                expires_at: future.to_rfc3339(),
                created_at: now.to_rfc3339(),
                updated_at: now.to_rfc3339(),
            });
        }

        cache.evict_expired();
        assert_eq!(cache.data.read().unwrap().len(), 1);
        assert!(cache.data.read().unwrap().contains_key("valid"));
    }

    #[test]
    fn evict_expired_no_expired_does_nothing() {
        let cache = CacheStore::new();
        let now = Utc::now();
        let future = now + Duration::hours(1);

        {
            let mut data = cache.data.write().unwrap();
            data.insert("valid".to_string(), CacheEntry {
                key: "valid".to_string(),
                value: "fresh".to_string(),
                ttl: 3600,
                expires_at: future.to_rfc3339(),
                created_at: now.to_rfc3339(),
                updated_at: now.to_rfc3339(),
            });
        }

        cache.evict_expired();
        assert_eq!(cache.data.read().unwrap().len(), 1);
    }

    #[test]
    fn stats_counts_expired_and_total_entries() {
        let cache = CacheStore::new();
        let now = Utc::now();
        let past = now - Duration::hours(1);
        let future = now + Duration::hours(1);

        {
            let mut data = cache.data.write().unwrap();
            data.insert("expired".to_string(), CacheEntry {
                key: "expired".to_string(),
                value: "old".to_string(),
                ttl: 3600,
                expires_at: past.to_rfc3339(),
                created_at: past.to_rfc3339(),
                updated_at: past.to_rfc3339(),
            });
            data.insert("valid".to_string(), CacheEntry {
                key: "valid".to_string(),
                value: "fresh".to_string(),
                ttl: 3600,
                expires_at: future.to_rfc3339(),
                created_at: now.to_rfc3339(),
                updated_at: now.to_rfc3339(),
            });
        }

        let stats = cache.stats();
        assert_eq!(stats["total_entries"], 2);
        assert_eq!(stats["expired_entries"], 1);
    }
}
